gpflow/multioutput/features.py

```diff
@@ -0,0 +1,150 @@
+        return len(self.feat)
+
+
+class SeparateIndependentMof(Mof):
+    """
+    A different feature is used for each output.
+    Note: each feature should have the same number of points, M.
+    """
+    def __init__(self, feat_list):
+        Mof.__init__(self)
+        self.feat_list = feat_list
+
+    def __len__(self):
+        return len(self.feat_list[0])
+
+
+class MixedKernelSharedMof(SharedIndependentMof):
+    """
+    This Mof is used in combination with the `SeparateMixedMok`.
+    Using this feature with the `SeparateMixedMok` leads to the most efficient code.
+    """
+    pass
+
+
+# ---------------------------------------------------------------------
+################################ Kuf ##################################
+# ---------------------------------------------------------------------
+
+@dispatch(InducingPoints, Mok, object)
+def Kuf(feat, kern, Xnew):
+    return kern.K(feat.Z, Xnew, full_cov_output=True)  #  M x P x N x P
+
+
+@dispatch(SharedIndependentMof, SharedIndependentMok, object)
+def Kuf(feat, kern, Xnew):
+    return Kuf(feat.feat, kern.kern, Xnew)  # M x N
+
+
+@dispatch(SeparateIndependentMof, SharedIndependentMok, object)
+def Kuf(feat, kern, Xnew):
+    return tf.stack([Kuf(f, kern.kern, Xnew) for f in feat.feat_list], axis=0)  # L x M x N
+
+
+@dispatch(SharedIndependentMof, SeparateIndependentMok, object)
+def Kuf(feat, kern, Xnew):
+    return tf.stack([Kuf(feat.feat, k, Xnew) for k in kern.kern_list], axis=0)  # L x M x N
+
+
+@dispatch(SeparateIndependentMof, SeparateIndependentMok, object)
+def Kuf(feat, kern, Xnew):
+    return tf.stack([Kuf(f, k, Xnew) for f, k in zip(feat.feat_list, kern.kern_list)], axis=0)  # L x M x N
+
+
+@dispatch((SeparateIndependentMof, SharedIndependentMof), SeparateMixedMok, object) 
+def Kuf(feat, kern, Xnew):
+    kuf_impl = Kuf.dispatch(type(feat), SeparateIndependentMok, object)
+    K = tf.transpose(kuf_impl(feat, kern, Xnew), [1, 0, 2])  # M x L x N
+    return K[:, :, :, None] * tf.transpose(kern.W)[None, :, None, :]  # M x L x N x P
+
+
+@dispatch(MixedKernelSharedMof, SeparateMixedMok, object)
+def Kuf(feat, kern, Xnew):
+    return tf.stack([Kuf(feat.feat, k, Xnew) for k in kern.kern_list], axis=0)  # L x M x N
+
+
+# ---------------------------------------------------------------------
+################################ Kuu ##################################
+# ---------------------------------------------------------------------
+
+@dispatch(InducingPoints, Mok)
+def Kuu(feat, kern, *, jitter=0.0):
+    Kmm = kern.K(feat.Z, full_cov_output=True)  # M x P x M x P
+    M = tf.shape(Kmm)[0] * tf.shape(Kmm)[1]
+    jittermat = jitter * tf.reshape(tf.eye(M, dtype=float_type), tf.shape(Kmm))
+    return Kmm + jittermat
+
+
+@dispatch(SharedIndependentMof, SharedIndependentMok)
+def Kuu(feat, kern, *, jitter=0.0):
+    Kmm = Kuu(feat.feat, kern.kern)  # M x M
+    jittermat = tf.eye(len(feat), dtype=float_type) * jitter
+    return Kmm + jittermat
```

Comment:
This seems to adding jitter twice since Kuu already add a jitter? I am not sure about all cases but at least for some cases it adds jitter twice.

Reply:
Maybe I'm missing something, but I don't see where/why the jitter is added twice? The first time `Kuu` is called without `jitter` arguments, which means that the default (i.e. `0`) is used. We then add `jittermat` once. The same goes for the other cases.

Reply:
You're right. I think I was confused: sorry!